pkg/os/shell/shell.go

```diff
@@ -81,8 +90,137 @@ func GetPathEnvString(userShell string, prependedPath string) string {
+	}
+	return path
+}
+
+// convertToWindowsSubsystemLinuxPath is a utility method to translate between Windows and WSL(Windows Subsystem for
+// Linux) paths. It relies on `wslpath` command to perform this conversion.
+//
+// It receives one argument:
+// - path : Windows path to be converted to WSL path
+//
+// It returns translated WSL equivalent of provided windows path.
+func convertToWindowsSubsystemLinuxPath(path string) string {
+	stdOut, _, err := CommandRunner.Run("wsl", "-e", "bash", "-c", fmt.Sprintf("wslpath -a '%s'", path))
+	if err != nil {
+		return path
+	}
+	return strings.TrimSpace(stdOut)
+}
+
+// IsWindowsSubsystemLinux detects whether current system is using Windows Subsystem for Linux or not
+//
+// It checks for these conditions to make sure that current system has WSL installed:
+// - `/proc/version` file is present
+// - `/proc/version` file contents contain keywords `Microsoft` and `WSL`
+//
+// It above conditions are met, then this method returns `true` otherwise `false`.
+func IsWindowsSubsystemLinux() bool {
+	procVersionContent, err := os.ReadFile(WindowsSubsystemLinuxKernelMetadataFile)
+	if err != nil {
+		return false
+	}
+	if strings.Contains(string(procVersionContent), "Microsoft") ||
+		strings.Contains(string(procVersionContent), "WSL") {
+		return true
+	}
+	return false
+}
+
+// detectShellByInvokingCommand is a utility method that tries to detect current shell in use by invoking `ps` command.
+// This method is extracted so that it could be used by unix systems as well as Windows (in case of WSL). It executes
+// the command provided in the method arguments and then passes the output to inspectProcessOutputForRecentlyUsedShell
+// for evaluation.
+//
+// It receives two arguments:
+// - defaultShell : default shell to revert back to in case it's unable to detect.
+// - command: command to be executed
+// - args: a string array containing command arguments
+//
+// It returns a string value representing current shell.
+func detectShellByInvokingCommand(defaultShell string, command string, args []string) string {
+	stdOut, _, err := CommandRunner.Run(command, args...)
+	if err != nil {
+		return defaultShell
+	}
+
+	detectedShell := inspectProcessOutputForRecentlyUsedShell(stdOut)
+	if detectedShell == "" {
+		return defaultShell
+	}
+	return detectedShell
+}
+
+// inspectProcessOutputForRecentlyUsedShell inspects output of ps command to detect currently active shell session.
+//
+// It takes one argument:
+//
+// - psCommandOutput: output of ps command executed on a particular shell session
+//
+// It returns:
+//
+//   - a string value (one of `zsh`, `bash` or `fish`) for current shell environment in use. If it's not able to determine
+//     underlying shell type, it returns and empty string.
+//
+// This method tries to check all processes open and filters out shell sessions (one of `zsh`, `bash` or `fish)
+// It then tries to sort these processes by process ids and returns the most recent one.
```

Comment:
In this method you are not performing any sorting because the `ps` command which is used already performing the sort with pid so better to update that info. https://play.golang.com/p/QhNeqa5tnEO

Reply:
@praveenkumar : Sorry, I forgot to update the method comment after making changes to the method.

Should be okay now.